bson/src/main/org/bson/internal/CodecCache.java

```diff
@@ -19,36 +19,75 @@
 import org.bson.codecs.Codec;
 import org.bson.codecs.configuration.CodecConfigurationException;
 
+import java.lang.reflect.Type;
+import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 import java.util.concurrent.ConcurrentHashMap;
 import java.util.concurrent.ConcurrentMap;
 
 import static java.lang.String.format;
 
 final class CodecCache {
-    private final ConcurrentMap<Class<?>, Optional<Codec<?>>> codecCache = new ConcurrentHashMap<>();
 
-    public boolean containsKey(final Class<?> clazz) {
-        return codecCache.containsKey(clazz);
+    static final class CodecCacheKey {
+        private final Class<?> clazz;
+        private final List<Type> types;
+
+        CodecCacheKey(final Class<?> clazz, final List<Type> types) {
```

Comment:
This seems like one more reason to introduce a new artifact.

Reply:
Not really, as we can't split a package between two modules like that.

Reply:
I did not notice at first: both `TypeWithTypeParameters` and `TypeData` are in `bson`, i.e., are already available to `bson-record-codec` and to `CodecRegistry`, as it is also in `bson`.

Reply:
Oh, of course...  Still, it would be weird if CodecRegistry depended on an interface in `org.bson.codecs.pojo`.  I also don't know if that's exactly the interface we want anyway.